NiFiDeploy.groovy

```diff
@@ -67,6 +67,94 @@ if (opts.debug) {
 
 // implementation methods below
 
+def handleGracefullShutdown() {
+  if (!conf.nifi.gracefullShutdown){
+    return
+  }
+
+  println "Attempting to gracefully shut down the pipeline"
+
+  // get the processGroups that have 0 incoming connections
+  // and add to ArrayList
+  def checkProcessGroups = ['root']
+  processGroups.each { pcGroup ->
+    if (pcGroup.inputPortCount == 0){
+      checkProcessGroups.add(pcGroup.id)
+    }
+  }
+
+  // get all processors in the found processGroups and add them to hashmap
+  // processorConnectionMap. This will be used to count the number of
+  // destinations
+  // the processorPgMap hashmap will be used to lookup processorIds with
+  // processGroupIds
+  def processorConnectionMap = [:]
+  def processorPgMap = [:]
+  checkProcessGroups.each { pcGroupId ->
+    def resp = nifi.get(
+      path : "controller/process-groups/$pcGroupId/processors"
+    )
+    assert resp.status == 200
+    processors = resp.data.processors
+  }
+  processors.each { processor ->
+    processorConnectionMap.put(processor.id,0)
+    processorPgMap.put(processor.id, processor.parentGroupId)
+  }
+
+  // get all connections in the found processGroups
+  def connections = ""
+  checkProcessGroups.each { pcGroupId ->
+    def resp = nifi.get(
+      path : "controller/process-groups/$pcGroupId/connections"
+    )
+    assert resp.status == 200
+    connections = resp.data.connections
+  }
```

Comment:
} should be outside of connections loop to count for connections of each process groups